client/subscriptions.go

```diff
@@ -0,0 +1,137 @@
+	Error error
+}
+
+type Subscription struct {
+	client  *Client
+	id      string
+	inner   api.Streams_ReadClient
+	channel chan (chan *SubscriptionEvent)
+}
+
+func NewSubscription(client *Client, inner api.Streams_ReadClient, id string) *Subscription {
+	channel := make(chan (chan *SubscriptionEvent))
+
+	// It is not safe to consume a stream in different goroutines. This is why we only consume
+	// the stream in a dedicated goroutine.
+	//
+	// Current implementation doesn't terminate the goroutine. When a subscription is dropped,
+	// we keep user requests coming but will always send back a subscription dropped event.
+	// This implementation is simple to maintain while letting the user sharing their subscription
+	// among as many goroutines as they want.
+	go func() {
+		closed := false
+
+		for {
+			req := <-channel
+
+			if closed {
+				req <- &SubscriptionEvent{
+					Dropped: &SubscriptionDropped{
+						Error: fmt.Errorf("subscription has been dropped"),
+					},
+				}
+
+				continue
+			}
+
+			result, err := inner.Recv()
+			if err != nil {
+				log.Printf("[error] subscription has dropped. Reason: %v", err)
+
+				dropped := SubscriptionDropped{
+					Error: err,
+				}
+
+				req <- &SubscriptionEvent{
+					Dropped: &dropped,
+				}
+
+				closed = true
+
+				continue
+			}
+
+			switch result.Content.(type) {
+			case *api.ReadResp_Checkpoint_:
+				{
+					checkpoint := result.GetCheckpoint()
+					position := position.Position{
+						Commit:  checkpoint.CommitPosition,
+						Prepare: checkpoint.PreparePosition,
+					}
+
+					req <- &SubscriptionEvent{
+						CheckPointReached: &position,
+					}
+				}
+			case *api.ReadResp_Event:
+				{
+					event := result.GetEvent()
+					recordedEvent := event.GetEvent()
+					streamIdentifier := recordedEvent.GetStreamIdentifier()
+					eventAppeared := messages.RecordedEvent{
+						EventID:        protoutils.EventIDFromProto(recordedEvent),
+						EventType:      recordedEvent.Metadata[system_metadata.SystemMetadataKeysType],
+						ContentType:    protoutils.GetContentTypeFromProto(recordedEvent),
+						StreamID:       string(streamIdentifier.StreamName),
+						EventNumber:    recordedEvent.GetStreamRevision(),
+						CreatedDate:    protoutils.CreatedFromProto(recordedEvent),
+						Position:       protoutils.PositionFromProto(recordedEvent),
+						Data:           recordedEvent.GetData(),
+						SystemMetadata: recordedEvent.GetMetadata(),
+						UserMetadata:   recordedEvent.GetCustomMetadata(),
+					}
+
+					req <- &SubscriptionEvent{
+						EventAppeared: &eventAppeared,
+					}
+				}
+			}
+		}
+	}()
+
+	return &Subscription{
+		client:  client,
+		id:      id,
+		inner:   inner,
+		channel: channel,
+	}
+}
+
+func (sub *Subscription) Id() string {
+	return sub.id
+}
+
+func (sub *Subscription) Close() error {
+	return sub.inner.CloseSend()
+}
+
+func (sub *Subscription) Recv() *SubscriptionEvent {
+	req := make(chan *SubscriptionEvent)
+
+	sub.channel <- req
```

Comment:
No idea how much it affects performance, but maybe it's an idea to look into using a single channel.

Reply:
Then I don't khow to make that possible while keep using channels. An alternative implementation would be to use a mutex around the subscription stream.